feat: upgrade to Lean 4.5.0

monsterkrampe opened the pull request:
Hi!

I wanted to play around with the library and it took me some time to make it build locally.
I also wanted to see if it is possible to upgrade to the current stable lean (and mathlib) version.
In the process I understood why you went for forking mathlib to have a slightly different definition for `MvQPF` but I got it working somehow :sweat_smile: 
I create this PR since I want to save others some of the trouble. Also, I changed the mathlib imports to not build all of mathlib (which ended up taking a very long time on my machine).

I should say that I am not a seasoned Lean developer (yet), so don't be shy to point out my mistakes :)
Also, unfortunately not everything is working again yet. One test in `Test/Misc` fails and I did not fully understand yet why this happens. I think it worked before upgrading to 4.5.0. Maybe I screwed up something in the macros... 
If you have any ideas, I would really appreciate your help!

Also, I'm curious what your plan for this library is. As far as I understood, this was mainly your master thesis project. Since I feel that coinduction support in Lean is somewhat lacking, I would love to see this library getting more traction.

Cheers and thanks for your work! :)

alexkeizer commented:
Thanks for your PR! I've been a bit preoccupied with other projects, but I do intend to at least maintain this library, keeping it up to date (so your PR is very much appreciated!), and hopefully extending it to become a bit more user friendly.

> Also, I changed the mathlib imports to not build all of mathlib (which ended up taking a very long time on my machine).

In case you're not aware, if you run `lake exe cache get`, you'll download a cached build of mathlib, saving you the trouble of building it yourself. That said, minimizing the imports is still a good idea.

> I should say that I am not a seasoned Lean developer (yet), so don't be shy to point out my mistakes :)

Don't worry about that, I certainly wouldn't have called myself a seasoned Lean person when I started this project, so feel free to point out when you notice some parts of the library look off as well.

> Also, I'm curious what your plan for this library is. As far as I understood, this was mainly your master thesis project.
Indeed it was. I haven't really kept it up to date because I've been working on other projects. But, I do very much want to develop this to the point where it can actually be used, not in the least because I foresee a use for nice coinduction in those other projects I mentioned.

alexkeizer commented on Qpf/Qpf/Multivariate/Basic.lean:
```diff
@@ -1,15 +1,115 @@
+      MvFunctor <| TypeFun.ofCurried <| F.curried := 
+    by rw[TypeFun.ofCurried_curried_involution]; exact f
+
+  instance instQPF_ofCurried_curried [func : MvFunctor F] [q : MvQPF F] : 
+      MvQPF <| TypeFun.ofCurried <| F.curried :=
+    by 
+      have : TypeFun.ofCurried (TypeFun.curried F) = F ∧ HEq (@instMvFunctor_ofCurried_curried n F func) func := by 
+        constructor 
+        exact TypeFun.ofCurried_curried_involution
+        simp [instMvFunctor_ofCurried_curried]
+      rw[helperQpf this]; exact q
```
```suggestion
instance instQPF_ofCurried_curried [func : MvFunctor F] [q : MvQPF F] :
      MvQPF <| TypeFun.ofCurried <| F.curried :=
    by
      apply cast ?_ q
      congr
      · simp only [TypeFun.ofCurried_curried_involution]
      · exact (cast_heq _ _).symm
```

alexkeizer commented on Qpf/Qpf/Multivariate/Basic.lean:
```diff
@@ -1,15 +1,115 @@
       IsPolynomial <| TypeFun.ofCurried <| F.curried := 
-    by rw[TypeFun.ofCurried_curried_involution]; exact p
+    by 
+      have : TypeFun.ofCurried (TypeFun.curried F) = F ∧ HEq (@instMvFunctor_ofCurried_curried n F func) func := by 
+        constructor 
+        exact TypeFun.ofCurried_curried_involution
+        simp [instMvFunctor_ofCurried_curried]
+      rw[helperPoly this]; exact p
```
```suggestion
      apply cast ?_ p
      congr
      · simp only [TypeFun.ofCurried_curried_involution]
      · exact (cast_heq _ _).symm
```

alexkeizer commented on Qpf/Qpf/Multivariate/Basic.lean:
```diff
@@ -1,15 +1,115 @@
-      MvQPF <| TypeFun.ofCurried <| F.curried := 
-    by rw[TypeFun.ofCurried_curried_involution]; exact q
+  lemma helperFunc {A} {B} : A = B -> @MvFunctor n A = @MvFunctor n B := by
+    intro h
+    rw [h]
+
+  lemma helperQpf {A} {B} [func : @MvFunctor n A] [func' : @MvFunctor n B] : A = B ∧ HEq func func' -> MvQPF A = MvQPF B := by 
+    intro h 
+    congr
+    exact h.left
+    exact h.right
 
-  instance instIsPolynomial_ofCurried_curried [p : IsPolynomial F] : 
+  lemma helperPoly {A} {B} [func : @MvFunctor n A] [func' : @MvFunctor n B] : A = B ∧ HEq func func' -> IsPolynomial A = IsPolynomial B := by 
+    intro h 
+    congr
+    exact h.left
+    exact h.right
```
Please remove these lemmas (see my suggestions below)

alexkeizer commented on Qpf/Qpf/Multivariate/ofPolynomial.lean:
```diff
@@ -11,27 +11,39 @@ import Qpf.Macro.Tactic.FinDestr
 -/
 
 namespace MvQPF
+  def ofIsomorphismMvFunctor {F : TypeFun n} 
+                    (F' : TypeFun n)
+                    [q : MvFunctor F']
+                    (box    : ∀{α}, F α → F' α) 
+                    (unbox  : ∀{α}, F' α → F α) 
+                  : MvFunctor F
+    where
+      map f a     := unbox <| q.map f <| box a
+
```
```suggestion
namespace MvFunctor
  def ofIsomorphism {F : TypeFun n} 
                    (F' : TypeFun n)
                    [q : MvFunctor F']
                    (box    : ∀{α}, F α → F' α) 
                    (unbox  : ∀{α}, F' α → F α) 
                  : MvFunctor F
    where
      map f a     := unbox <| q.map f <| box a

end MvFunctor

end MvQpf
```

Nice idea to split this out!

alexkeizer commented on Qpf/Qpf/Multivariate/ofPolynomial.lean:
```diff
@@ -11,27 +11,39 @@ import Qpf.Macro.Tactic.FinDestr
                     (unbox_box_id : ∀{α} (x : F α), unbox (box x) = x 
                                   := by intros; rfl
                                 )
+                    (func : MvFunctor F)
+                    (func_map : ∀ (α β : TypeVec n) (f : TypeVec.Arrow α β) (a : F α), func.map f a = (unbox <| func'.map f <| box a))
```
Why do we need these extra arguments?

monsterkrampe replied:
To be able to return `MvQPF F`, we need to make sure that we have `MvFunctor F`. To make `abs_map` work, we want to ensure that this instance is particularly the one that we get from `MvFunctor.ofIsomorphism` above. I decided to express this assertion by passing a prop for the necessary equality. Maybe there is a better way of achieving this :)

alexkeizer replied:
Ah, right. This used to be bundled, but now that we're using normal Mathlib again, we need the instance to be able to write `MvQPF F`. We could avoid this, by passing the instance explicitly, as in `@MvQPF _ F (MvFunctor.ofIsomorphism ..)`, but probably it's nicer to allow the possibility of a different instance as well.

However, I would suggest we make have the instance be inferred, and use an auto-tactic for the proof. This way, in the common case we don't actually have to supply anything.
```suggestion
                    [functor : MvFunctor F]
                    (map_eq : ∀ (α β : TypeVec n) (f : TypeVec.Arrow α β) (a : F α), functor.map f a = (unbox <| func'.map f <| box a) := by intros; rfl)
```

monsterkrampe replied:
I'll give it a shot :) 
I'll also give proper names in places where I used `func`. I did this all pretty quick and dirty for now and did not adjust afterwards. Sorry for that :D

monsterkrampe replied:
Alright, I did the adjustment, also in `Qpf/Qpf/Multivariate/Basic.lean`. Everything seems to work with the implicit argument now :)

alexkeizer commented on Qpf/Util/Vec.lean:
```diff
@@ -202,7 +206,9 @@ namespace Vec
     case nil          => rfl
-    case cons a as ih => simp only [toList, ofList, append1, last, DVec.last, drop, ih]
+    case cons a as ih => 
+      have : drop (append1 (ofList as) a) = ofList as := by unfold drop append1; dsimp
+      simp only [toList, ofList, append1, last, DVec.last, this, ih]
```
```suggestion
      simp only [toList, ofList, append1, last, DVec.last, drop_append1', ih]
```

alexkeizer commented on Test/Misc.lean:
```diff
@@ -22,8 +22,9 @@ data QpfList₃ α where
 
-data QpfTest α β where
-  | A : α → α → β → QpfTest α β → QpfTree β → QpfCoTree (QpfTree (QpfTest α β)) → QpfTest α β
+-- FIXME: this broke after Upgrading to Lean/Mathlib 4.5.0 and the associated changes
+-- data QpfTest α β where
+--   | A : α → α → β → QpfTest α β → QpfTree β → QpfCoTree (QpfTree (QpfTest α β)) → QpfTest α β
```
I filed an issue (#2) with my thoughts on this issue, and how to solve it.
I'm fine with merging this and implementing the fix later, so don't worry about it

monsterkrampe replied:
Sounds good, thanks! :)

alexkeizer reviewed (commented):
Thanks for the PR!

alexkeizer commented on Qpf/Macro/Comp.lean:
```diff
@@ -71,13 +79,14 @@ where
 
     trace[QPF] "F := {F}\nargs := {args.toList}\ndepth := {depth}"
-    QQ.check F
+    check F
```
```suggestion
    F.check
```

The function `QQ.check` was moved to `QQ.Quoted.check`. Your suggestion would use `Lean.Meta.check`, which is slightly different: the `Lean` function checks that the expression is well-typed (for some type), whereas the `Qq` version additionally checks that the type corresponds to the declared type (i.e., `CurriedTypeFun.{u, u} $depth`)

monsterkrampe replied:
I wasn't aware of that, thanks!

monsterkrampe commented:
Thanks for the quick reply and the review!  
I remember coming across `lake exe cache get` but I did not consider it here. Thank you for the reminder :)

I love to hear that you still plan to push development this repo. I totally get that time is short being involved in other projects. I'm happy to help wherever I can (although I still need to understand things better I think).

I addressed most of your comments now and I'm about to answer some of the other questions inline :)

monsterkrampe commented on Qpf/Macro/Comp.lean:
```diff
@@ -317,6 +327,11 @@ def elabQpfComposition (view: QpfCompositionView) : CommandElabM Unit := do
+      $modifiers:declModifiers
+      instance : 
+        MvFunctor ($F_internal_applied) := 
+      by unfold $F_internal; infer_instance
```
For now, I just ended up adding this because otherwise some of the tests would fail.
Which other instance do you mean in particular? I'm afraid I'm not familiar enough with the code yet to see the connection :sweat_smile:

alexkeizer replied:
Ah, I'd misread the diff. I was referring to the following lines, not realizing that GitHub was omitting the few hundred lines in between.
```lean
let func ← synthMvFunctor F
let _ ← synthQPF F func
```

Please ignore my comment

alexkeizer commented:
Looks good! Thanks for making the changes
